src/Aspire.Hosting.Azure.KeyVault/AzureKeyVaultResourceExtensions.cs

```diff
@@ -32,23 +34,52 @@ public static IResourceBuilder<AzureKeyVaultResource> AddAzureKeyVault(this IDis
+    /// <param name="configureResource">Callback to configure the underlying <see cref="global::Azure.Provisioning.KeyVault.KeyVaultService"/> resource.</param>
     /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
     [Experimental("AZPROVISION001", UrlFormat = "https://aka.ms/dotnet/aspire/diagnostics#{0}")]
-    public static IResourceBuilder<AzureKeyVaultResource> AddAzureKeyVault(this IDistributedApplicationBuilder builder, string name, Action<IResourceBuilder<AzureKeyVaultResource>, ResourceModuleConstruct, KeyVault>? configureResource)
+    public static IResourceBuilder<AzureKeyVaultResource> AddAzureKeyVault(this IDistributedApplicationBuilder builder, string name, Action<IResourceBuilder<AzureKeyVaultResource>, ResourceModuleConstruct, KeyVaultService>? configureResource)
     {
         builder.AddAzureProvisioning();
 
         var configureConstruct = (ResourceModuleConstruct construct) =>
         {
-            var keyVault = construct.AddKeyVault(name: construct.Resource.Name);
-            keyVault.AddOutput("vaultUri", x => x.Properties.VaultUri);
+            var keyVault = new KeyVaultService(construct.Resource.Name)
+            {
+                Location = construct.AddLocationParameter(),
+                Properties = new KeyVaultProperties()
+                {
+                    TenantId = new MemberExpression(new FunctionCallExpression(new IdentifierExpression("tenant")), "tenantId"),
+                    Sku = new KeyVaultSku()
+                    {
+                        Family = KeyVaultSkuFamily.A,
+                        Name = KeyVaultSkuName.Standard
+                    },
+                    EnableRbacAuthorization = true
+                }
+            };
+            construct.Add(keyVault);
 
-            keyVault.Properties.Tags["aspire-resource-name"] = construct.Resource.Name;
+            construct.Add(new BicepOutput("vaultUri", typeof(string))
+            {
+                Value =
+                    new MemberExpression(
+                        new MemberExpression(
+                            new IdentifierExpression(keyVault.ResourceName),
+                            "properties"),
+                        "vaultUri")
+            });
 
-            var keyVaultAdministratorRoleAssignment = keyVault.AssignRole(RoleDefinition.KeyVaultAdministrator);
-            keyVaultAdministratorRoleAssignment.AssignProperty(x => x.PrincipalId, construct.PrincipalIdParameter);
-            keyVaultAdministratorRoleAssignment.AssignProperty(x => x.PrincipalType, construct.PrincipalTypeParameter);
+            keyVault.Tags["aspire-resource-name"] = construct.Resource.Name;
+
+            var role = KeyVaultBuiltInRole.KeyVaultAdministrator;
+            var roleAssignment = new RoleAssignment($"{KeyVaultBuiltInRole.GetBuiltInRoleName(role)}_{keyVault.ResourceName}")
```

Comment:
Does GetBuiltInRole have an interface (static interface method). We should be able to make this method generic.

Reply:
You should be able to punt on this with the new generated `AssignRole` overloads.

Reply:
> You should be able to punt on this with the new generated AssignRole overloads.

Yep, for all except WebPubSub, which doesn't have a BuiltInRole.

Reply:
I just need to add some metadata to the spec (i.e., like [this example](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/provisioning/Generator/src/Specifications/SignalRSpecification.cs#L29-L35)) but didn't see anything documented at https://learn.microsoft.com/en-us/azure/role-based-access-control/built-in-roles other than `SignalR/Web PubSub Contributor`.  I was going to follow-up with Liangying, but David beat me to it here.  😆  Are https://github.com/microsoft/azure-roles/blob/main/azure_roles.json#L585-L586 sufficient to add?

Reply:
How about https://github.com/Azure/azure-sdk-for-net/pull/45977 ? For some reason that the team is now investigating, the builtin roles for Web PubSub are listed in Preview category somehow, so that they are not seen in the documentation.